core/src/apps/solana/sign_tx.py

```diff
@@ -167,4 +185,136 @@ def calculate_fee(transaction: Transaction) -> int:
+                instruction.parsed_data["space"] + SOLANA_ACCOUNT_METADATA_SIZE
+            )
+        elif instruction.program_id == _ASSOCIATED_TOKEN_ACCOUNT_PROGRAM_ID and (
+            instruction.instruction_id
+            == _ASSOCIATED_TOKEN_ACCOUNT_PROGRAM_ID_INS_CREATE
+            or instruction.instruction_id
+            == _ASSOCIATED_TOKEN_ACCOUNT_PROGRAM_ID_INS_CREATE_IDEMPOTENT
+        ):
+            spl_token_account = lookup_account(
+                transaction, instruction.parsed_accounts["spl_token"]
+            )
+            if spl_token_account == base58.decode(_TOKEN_PROGRAM_ID):
+                allocation_estimate += (
+                    SOLANA_TOKEN_ACCOUNT_SIZE + SOLANA_ACCOUNT_METADATA_SIZE
+                )
+            elif spl_token_account == base58.decode(_TOKEN_2022_PROGRAM_ID):
+                # https://docs.rs/spl-token-2022/latest/spl_token_2022/extension/enum.ExtensionType.html
+                # We only count the account extensions towards the allocation
+                allocation_estimate += (
+                    SOLANA_TOKEN_ACCOUNT_SIZE
+                    + (
+                        SOLANA_TVL_HEADER_SIZE
+                        + SOLANA_TOKEN22_TRANSFER_FEE_AMOUNT_EXTENSION_SIZE
+                    )
+                    + (
+                        SOLANA_TVL_HEADER_SIZE
+                        + SOLANA_TOKEN22_CONFIDENTIAL_TRANSFER_ACCOUNT_EXTENSION_SIZE
+                    )
+                    + (
+                        SOLANA_TVL_HEADER_SIZE
+                        + SOLANA_TOKEN22_IMMUTABLE_OWNER_EXTENSION_SIZE
+                    )
+                    + (
+                        SOLANA_TVL_HEADER_SIZE
+                        + SOLANA_TOKEN22_MEMO_TRANSFER_EXTENSION_SIZE
+                    )
+                    + (SOLANA_TVL_HEADER_SIZE + SOLANA_TOKEN22_CPI_GUARD_EXTENSION_SIZE)
+                    + (
+                        SOLANA_TVL_HEADER_SIZE
+                        + SOLANA_TOKEN22_NON_TRANSFERABLE_ACCOUNT_EXTENSION_SIZE
+                    )
+                    + (
+                        SOLANA_TVL_HEADER_SIZE
+                        + SOLANA_TOKEN22_TRANSFER_HOOK_ACCOUNT_EXTENSION_SIZE
+                    )
+                    + (
+                        SOLANA_TVL_HEADER_SIZE
+                        + SOLANA_TOKEN22_CONFIDENTIAL_TRANSFER_FEE_ACCOUNT_EXTENSION_SIZE
+                    )
+                    + (
+                        SOLANA_TVL_HEADER_SIZE
+                        + SOLANA_TOKEN22_PAUSABLE_ACCOUNT_EXTENSION_SIZE
+                    )
+                    + SOLANA_ACCOUNT_METADATA_SIZE
+                )
+            else:
+                # REVIEW: Is this case (not a Token nor a Token22 account) actually possible?
+                ...
+
+    rent_exemption_estimate = (
+        allocation_estimate
+        * SOLANA_RENT_PER_BYTE_EPOCH
+        * SOLANA_RENT_EXEMPTION_MULTIPLIER
+    )
+    return rent_exemption_estimate
```

Comment:
Actually I think we should consider this an "estimate" only in the case a Token22 instruction is used. I.e. IMO this should return something like `return is_estimate, rent` and it should be used accordingly in other places.